### ISSUE TITLE
Add support for refactoring variables (requires language server changes)

### ISSUE THREAD

**Issue by bleuthoot-sven** [godotengine/godot-vscode-plugin#164]
When I want to refactor variables in GDScript, I get this error:

```
[Error - 21:01:32] Request textDocument/rename failed.
  Message: Method not found
  Code: -32601
```

![](https://i.imgur.com/A5SiijR.gif)

**Comment by Calinou** [comment]
Were refactoring tools ever implemented in the GDScript language server?

**Comment by Geequlim** [comment]
@Calinou No :(

**Comment by git2013vb** [comment]
Well, when it will be implemented? 
Its a very important step for every language. So for Godot too. 
It will improve tremendous the way to use gdscript. Someone heed to start this important task. I'm not an expert, otherwise I would start it by myself.

**Comment by atirut-w** [comment]
You mean [this](https://github.com/godotengine/godot/pull/80973)?

**Comment by git2013vb** [comment]
> You mean [this](https://github.com/godotengine/godot/pull/80973)?

Nope. I was referring some common tasks that refactoring do in other environments.
for example:

* Red-Green Refactoring
* Extract Method
* Simplifying Methods
* Composing Method
* Abstraction

source: https://www.cloudzero.com/blog/refactoring-techniques

If I can choose two among them to have a priority they would be : Extract Method and Simplify Methods

**Comment by atirut-w** [comment]
I was talking to OP, sorry. Kind of forgot to make that clear.

**Comment by DaelonSuzuka** [comment]
> Well, when it will be implemented? ... I'm not an expert, otherwise I would start it by myself.

@git2013vb Seems like you already have your answer. I wasn't an "expert" in TextMate Grammars or TypeScript or the VSCode Extension API, but I wanted to fix things so I went and figured it out.

**Comment by git2013vb** [comment]
> > Well, when it will be implemented? ... I'm not an expert, otherwise I would start it by myself.
> 
> @git2013vb Seems like you already have your answer. I wasn't an "expert" in TextMate Grammars or TypeScript or the VSCode Extension API, but I wanted to fix things so I went and figured it out.

Your right :) But if I start to learn how to implement it I'm quite sure I will not have time anymore to work in my projects. Also because I'm not familiar with extension lifecycle also with no experience I'm sure I will take long to get some results. Unless I will stop to sleep.. ;) In that case I think its doable.

**Comment by DaelonSuzuka** [comment]
Renaming variables has been fixed as of Godot v4.2.
